### ISSUE TITLE
terminado

### ISSUE THREAD

**Issue by asmodehn** [wsowens/sockterm#1]
Hi,

I was looking for a terminal in web frontend for an app I am planning to build and ended up here...
I'm fairly new to Elm, but I have a somewhat wide experience in programming.

I am interested in plugging this to a backend, like https://terminado.readthedocs.io/en/latest/ for instance.

I am trying to determine if it would be better to interface xterm.js with elm, and hook that up to a backend,
*OR* to hook up sockterm to terminado...

Have you already tried anything like that ? any hint/advice/opinion on that topic ?

**Comment by wsowens** [comment]
Hi, thanks for your interest in this project!

So just to be clear, you're planning on making the app with Elm to begin with? If so, I think using the underlying Elm package would be useful for you. Because of the whole "kernel code" controversy, I can't include the websocket functionality in the Elm package, but you only need a few ports and some boilerplate JavaScript to make it work. I guess whether you use xterm.js or this project, you'll be writing ports either way.

But short answer, yes, from looking at the documentation, I think this should work. In fact, this is exactly the sort of thing I had in mind when I wrote this.

Right now, only some of the ANSI escape codes are supported, so some of the functionality might not work. That said, this project is under active development. I'm open to accepting PRs and I'm eager to add features if I see interest.

You said that you're somewhat new to Elm, is there anything I can do to help you get started with this project?

**Comment by asmodehn** [comment]
Yes, I am planning for the frontend to be in elm, as I would need quite a lot of interactive behavior, and I want to stay protected from the javascript mess ;-) And there are so many things I'd rather work on other than web frontend, so the earlier it is working, and stable, the better...

I have seen wsowens/term, and although I like the idea of having a terminal entirely written in elm as a developer, I cannot afford to spend too much time on it (my focus is elsewhere), and I just want a working webbased terminal ;-)

So I see two aspects from the "usage" viewpoint:

- sockterm could have two webbased terminal frontends. Having an xterm.js frontend could help for functionality comparison with wsowens/term, and provide users with a "well-known" terminal implementation if they wish. These things can be soooo tricky to get right, thats why I want to stay away from their implementation if possible ;-)
But maybe it would be better suited for wsowens/term to provide an "elm wrapper" around xterm.js ? what do you think?

- sockterm could connect to various backend : my first interest here is terminado as I already have multiple python backends I could integrate it with.

To help get me started... hum, let's see... how about listing small bite-sized, testable, issues, that are baby steps along the way to these goals ?
How about deciding on a list of issues as "mini-stories", gather them in two milestones (in the appropriate project), and we can start working towards it ?

**Comment by wsowens** [comment]
> I want to stay protected from the javascript mess 

When Elm upgraded to 0.18, the [websocket](https://github.com/elm-lang/websocket) package was dropped. Furthermore, Elm 0.19 and onward forbids delivering packages with any JavaScript (so-called "kernel code", [read more here](https://github.com/elm-lang/websocket)). So no matter what, if you want to connect to `terminado` in Elm, you have to write a little bit of boilerplate JavaScript. (If you're unfamiliar with ports, you can [consult the Elm guide](https://guide.elm-lang.org/interop/ports.html).)

> sockterm could have two webbased terminal frontends.

`sockterm` is just a frontend. Implementation-wise, `sockterm` just an Elm [Term](https://package.elm-lang.org/packages/wsowens/term/latest/Term) hooked up to a websocket with some boilerplate.

I think providing a Elm wrapper around `xterm.js` is a cool idea. Unfortunately, that brings us back to the "Kernel Code" issue, since you can't distribute an Elm package with arbitrary JavaScript. So, if we did this, would write an Elm module with ports to interact with xterm.js, then have load xterm.js, initialize the Elm module, then hook up all the ports... all in JavaScript.
Based on your requirements, especially your desire to minimize the amount of JavaScript that you write, I   

> sockterm could connect to various backend : my first interest here is terminado as I already have multiple python backends I could integrate it with.

That's the entire idea! You should be able to do this *right now* if you just copy and paste [this example](https://github.com/wsowens/sockterm/blob/master/examples/basic.html).  I also have a live demo, but I need to tweak one thing to comply with the CORS policies. (I'll leave another comment in this thread in a few hours.)

**Comment by wsowens** [comment]
![image](https://user-images.githubusercontent.com/22205370/97965032-921e4f00-1d6e-11eb-9f3c-6e306adf2288.png)

I'm not familiar with Tornado, but it seems like `tornado.web.WebSocketHandler` expects you to have already established a connection to the server. So, you can't use my [live demo](http://term.ovvens.com) to connect to a tornado server, but that's ok, we can just serve a copy of the HTML inside the terminado server.

I wrote a very rough backend with terminado, just to show that it's possible. Note that I didn't touch the sockterm code at all, because sockterm is just a frontend. 

I took a few shortcuts for this proof-of-concept:
1) I sidestepped the JSON parsing for now. (That can *easily* be implemented in Elm.)
2) I manually filtered out the ANSI Escape Code for the OCR command. (I don't plan to implement this, it's low priority in my opinion.) However, I will update the Elm `term` package to NOT choke on this command.
3) I caught a bug with the color codes, but I just used a quick workaround.

I'll have 2 and 3 fixed shortly, but I wanted to show you where things stand.

**Comment by wsowens** [comment]
If you want to reproduce the screenshot above for yourself, you can follow these instructions:

1) Download [this example.zip file](https://github.com/wsowens/sockterm/files/5480235/example.zip) and unzip it somewhere.
2) Install the required modules (I recommend making a virtual environment first)
```
pip install -r requirements.txt
```
3) Launch the server
```
python example.py
```
4) Launch your browser and navigate to [localhost:8000](http://localhost:8000)
5) In the terminal's address bar, type `ws://localhost:8000/websocket` and press enter

You should be connected!

**Comment by asmodehn** [comment]
I understand your point about the elm wrapper around xterm.js, that might not be fit for the elm ecosystem at this time then.

Thanks a lot for the example code to connect to a terminado backend, I was able to run it.
I ll play with it for the next few days, to understand it better and try to get a "web-based terminal frontend" as I picture it for my usecase... I ll make a PR for another kind of example, when I get something working...

**Comment by wsowens** [comment]
Great! I'll fix the issues I encountered with the parser. I also have a few tweaks for the UI in mind as well.

**Comment by wsowens** [comment]
@asmodehn 

I have updated the Elm package to fix the two parser issues I found above. I also recompiled the files in `dist/` to reflect those changes.  Let me know if you have any questions about using the Elm package.

**Comment by asmodehn** [comment]
Is there a place in this repo for some backend code (to be used for tests or demo, maybe for color code, etc.) ??
I was thinking of a terminado / ipython combination to make a web-base python repl ...
Where should I put this ?

**Comment by wsowens** [comment]
Good call. How about `backend/`?

I actually have already written an ANSI color module for another project. (Check it out [here](https://github.com/ufosc/swampymud/blob/master/swampymud/util/color.py).) I can write up a simple websocket server and add it to the folder later this evening (~8 hours from now). 

Then someone could just connect to a websocket like  `ws://term.ovvens.com/colortest` and it would send them a ton of different ANSI-encoded SGR sequences.

**Comment by wsowens** [comment]
@asmodehn , I've added a backend `colortest.py` the `backend/` folder. I also have it running on my server at `ws://term.ovvens.com/colortest`.

Here's a screenshot of it running:
![image](https://user-images.githubusercontent.com/22205370/98759661-90bcda00-239f-11eb-8a1c-007dfce82e01.png)

Looking forward to seeing your terminado / ipython backend.

**Comment by wsowens** [comment]
Update: I made `colortest.py` a bit more interactive to facilitate debugging in the future.

![image](https://user-images.githubusercontent.com/22205370/98872055-f19dee00-2443-11eb-9925-5b7b1883986d.png)
